### PR TITLE
Redis tx callbacks

### PR DESCRIPTION
**Pull request by BernhardBln**
Provide a way for `@RedisTransactional` enabled subscribed projections to receive callbacks for commit and rollbacks of the redis transaction.

Enable on projection via `@EnableRedisTransactionCallbacks`.

Only works when projection is subclass of
`AbstractRedisSubscribedProjection`, and is annotated with `@RedisTransactional`.

Overwrite `onCommit` and `onRollback` to receive the notifications.

### CODE REVIEW

File: factcast-factus-redis/src/main/java/org/factcast/factus/redis/tx/EnableRedisTransactionCallbacks.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * <p>Attention, in case you are adding this to projections that are already deployed, it would be
+ * safer to increase the serial of the projection. Otherwise, during the first deployment, some
+ * updates might get lost when the instance that processes does not have this annotation, and the
+ * newly deployed instances do not have the lock but expect the callbacks.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Inherited
+public @interface EnableRedisTransactionCallbacks {}
```
Comment (uweschaefer) [inline factcast-factus-redis/src/main/java/org/factcast/factus/redis/tx/EnableRedisTransactionCallbacks.java:40]: obsolete, should be an interface

**Review by uweschaefer — COMMENTED** [review]
This is a good poc. However, we should not add this being specific to Redis. Looking at it, what the user wants is a callback on external data change. if we go the pulling way as a fallback, we can use the factPosition to detect such a change.
For Datastores that can do subscriptions, this should be overridden to use triggers from the datastore, like you demonstrated here.

File: factcast-factus-redis/src/main/java/org/factcast/factus/redis/tx/RedisTransactionalLens.java [hunk 2]
```diff
@@ -121,10 +144,16 @@ static long calculateFlushTimeout(@NonNull TransactionOptions opts) {
   @Override
   public void doClear() {
     txManagerSupplier.get().rollback();
```
[the remark this answers is not in the document]
Comment (BernhardBln) [reply factcast-factus-redis/src/main/java/org/factcast/factus/redis/tx/RedisTransactionalLens.java:146]: true, was more "I can, why not", but you are right.

File: factcast-factus-redis/src/main/java/org/factcast/factus/redis/AbstractRedisSubscribedProjection.java [hunk 3]
```diff
@@ -17,11 +17,72 @@
...
+                      .warn(
+                          "Received unknown code {} from channel {} in transaction notification subscription.",
+                          code,
+                          channel);
+              }
+            });
+  }
+
+  /**
+   * <b>Will only be called when the projection is annotated with <code>
```
[the remark this answers is not in the document]
Comment (BernhardBln) [reply factcast-factus-redis/src/main/java/org/factcast/factus/redis/AbstractRedisSubscribedProjection.java:68]: Initial idea was to limit it to subclasses of this projection, but if we decide to make it a general feature for every projection, I agree.

**Comment by BernhardBln** [comment]
> This is a good poc. However, we should not add this being specific to Redis. Looking at it, what the user wants is a callback on external data change. 

I kind of agree, but maybe it makes sense to think a little more what that means.

A projection, for example, can be local or distributed. For a local example this is still useful, it's like a "post event apply" method, but the term external data change makes less sense here.

And another thing, I kind of silently assumed that those subscribed redis projections, for which I made the initial change, would be subscribed to at app startup and for the whole lifetime of the application.

That might not necessarily be true at all times. So do we need a lifecycle for these notifications? Like, do not receive these events before actually subscribing, and stop receiving those events as soon as you stop subscribing? 
Is closing a subscription actually possible/desired? I understand you would do that by closing the factus instance?

That leads to other problems, in case of redis I have this little delay, like I change the fact position or flush the transaction, and that publishes something on a topic that needs to make a roundtrip to the server and back. If I was the one who had the lock and was processing, I still need to process this notification, but not the ones that come from another subscribed projection taking over?


What does this mean for Managed projections? Should they receive those updates or not?


My gut feeling is, it would make sense to limit this to subscribed projections (even if you lose the "post event apply" function for local projections), and the mechanism would start when you start the subscription, and what is still unclear to me is what happens for the end of the subscription and how we handle the problem I mentioned above.

**Comment by uweschaefer** [comment]
closed with #2342

**Pull request closed without merging**